core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,873 @@
+      },
+      question_state_data_schema_version: 45,
+      language_code: 'ab',
+      version: 1,
+      linked_skill_ids: [],
+      inapplicable_skill_misconception_ids: [],
+    }];
+
+    sampleExplorationDict = {
+      exploration_id: 'explorationId1',
+      is_logged_in: true,
+      session_id: 'KERH',
+      exploration: {
+        init_state_name: 'Introduction',
+        param_changes: [],
+        param_specs: null,
+        title: 'Exploration',
+        language_code: 'en',
+        correctness_feedback_enabled: true,
+        objective: 'To learn',
+        states: {
+          Introduction: {
+            param_changes: [],
+            classifier_model_id: null,
+            recorded_voiceovers: null,
+            solicit_answer_details: true,
+            card_is_checkpoint: true,
+            written_translations: null,
+            linked_skill_id: null,
+            next_content_id_index: null,
+            content: {
+              html: '',
+              content_id: 'content'
+            },
+            interaction: {
+              customization_args: {},
+              answer_groups: [],
+              solution: null,
+              hints: [],
+              default_outcome: {
+                param_changes: [],
+                dest: 'Introduction',
+                feedback: {
+                  html: '',
+                  content_id: 'content'
+                },
+                labelled_as_correct: true,
+                refresher_exploration_id: 'exp',
+                missing_prerequisite_skill_id: null
+              },
+              confirmed_unclassified_answers: [],
+              id: null
+            }
+          }
+        }
+      },
+      version: 1,
+      can_edit: true,
+      preferred_audio_language_code: 'en',
+      preferred_language_codes: [],
+      auto_tts_enabled: true,
+      correctness_feedback_enabled: true,
+      record_playthrough_probability: 1
+    };
+  });
+
+  beforeEach(() => {
+    alertsService = TestBed.inject(AlertsService);
+    answerClassificationService = TestBed.inject(AnswerClassificationService);
+    contextService = TestBed.inject(ContextService);
+    expressionInterpolationService =
+      TestBed.inject(ExpressionInterpolationService);
+    questionObjectFactory = TestBed.inject(QuestionObjectFactory);
+    readOnlyExplorationBackendApiService = TestBed.inject(
+      ReadOnlyExplorationBackendApiService);
+    questionPlayerEngineService = TestBed.inject(QuestionPlayerEngineService);
+    outcomeObjectFactory = TestBed.inject(OutcomeObjectFactory);
+    focusManagerService = TestBed.inject(FocusManagerService);
+    urlService = TestBed.inject(UrlService);
+  });
+
+  it('should load questions when initialized', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId')
```

Comment:
Why is this line needed (and get exploration version below)? Ditto in other test functions.

Reply:
removed all unnecessary spies too.